docs/architecture/adr-082-data-companion-push-api.md

```diff
@@ -0,0 +1,440 @@
+   (e.g. JSON-RPC or gRPC).
+2. Integrators can index data in whatever way suits them best.
+3. All users eventually benefit from faster changes to Tendermint because the
+   core team has less, and less complex, code to maintain (implementing this ADR
+   would add more code in the short-term, but would pave the way for significant
+   reductions in complexity in future).
+
+The way the system is currently built is such that a Tendermint node is mostly
+self-contained. While this philosophy initially allowed a certain degree of ease
+of node operation (i.e. simple UX), it has also lent itself to feature sprawl,
+with Tendermint being asked to take care of increasingly more than just
+consensus. Under the spotlight in this ADR are:
+
+1. **Event indexing**, which is required in order to facilitate arbitrary block
+   and transaction querying, as well as subscription for arbitrary events. We
+   have already seen, for instance, how the current (somewhat unreliable) event
+   subscription implementation can cause back-pressure into consensus and affect
+   IBC relayer stability (see [\#6729] and [\#7156]).
+2. **Block execution result storage**, which is required to facilitate a number
+   of RPC APIs, but also results in storage of large quantities of data that is
+   not critical to consensus. This data is, however, critical for certain use
+   cases outside of consensus.
+
+Another intersecting issue is that there may be a variety of use cases that
+require different data models. A good example of this is the mere existence of
+the [PostgreSQL indexer] as compared to the default key/value indexer.
+
+Continuing from and expanding on the ideas outlined in [RFC 006][rfc-006], the
+suggested approach in this ADR is to provide a mechanism to publish certain data
+from Tendermint, in real-time and with certain reliability guarantees, to a
+single companion service outside of the node that can use this data in whatever
+way it chooses (filter and republish it, store it, manipulate or enrich it,
+etc.).
+
+Specifically, this mechanism would initially publish:
+
+1. Block data (i.e. `Block` data structures that have been committed by the
+   consensus engine).
+2. The results of block execution (e.g. data from `FinalizeBlockResponse`). This
+   data is not accessible from the P2P layer, and currently provides valuable
+   information for Tendermint integrators (whether events should be handled at
+   all by Tendermint is a different problem).
+
+## Alternative Approaches
+
+1. One clear alternative to this would be the approach outlined in [ADR
+   075][adr-075]. This approach:
+
+   1. Still leaves Tendermint responsible for maintaining a query interface and
+      event indexing functionality, increasing the long-term maintenance burden of,
+      and the possibility of feature sprawl in, that subsystem. To overcome this,
+      we could remove the query interface altogether and just always publish all
+      data.
```

Comment:
As a user of the SDK, I'm personally pretty unconcerned about a Tendermint API for this, and am more than happy to do this in the SDK.

Not sure how other ABCI users feel.